x-pack/plugins/observability_solution/observability_onboarding/server/routes/flow/route.ts

```diff
@@ -171,8 +182,240 @@ const getProgressRoute = createObservabilityOnboardingServerRoute({
+    try {
+      agentInputs = await ensureInstalledIntegrations(integrationsToInstall, packageClient);
+    } catch (error) {
+      if (error instanceof FleetUnauthorizedError) {
+        return response.forbidden({
+          body: {
+            message: error.message,
+          },
+        });
+      }
+      throw error;
+    }
+
+    const authApiKey = getAuthenticationAPIKey(request);
+    const elasticsearchUrl = plugins.cloud?.setup?.elasticsearchUrl
+      ? [plugins.cloud?.setup?.elasticsearchUrl]
+      : await getFallbackESUrl(services.esLegacyConfigService);
+
+    return response.ok({
+      headers: {
+        'content-type': 'application/yaml',
+      },
+      body: generateAgentConfig({
+        apiKey: authApiKey ? `${authApiKey?.apiKeyId}:${authApiKey?.apiKey}` : '$API_KEY',
+        esHost: elasticsearchUrl,
+        inputs: agentInputs,
+      }),
+    });
+  },
+});
+
+type Integration =
+  | {
+      pkgName: string;
+      installSource: 'registry';
+    }
+  | {
+      pkgName: string;
+      installSource: 'custom';
+      logFilePaths: string[];
+    };
+
+async function ensureInstalledIntegrations(
+  integrationsToInstall: Integration[],
+  packageClient: PackageClient
+) {
+  const agentInputs: unknown[] = [];
+  for (const integration of integrationsToInstall) {
+    const { pkgName, installSource } = integration;
+    if (installSource === 'registry') {
+      await packageClient.ensureInstalledPackage({ pkgName });
+      agentInputs.push(...getSystemLogsDataStreams(uuidv4()));
+    } else if (installSource === 'custom') {
+      const input = {
+        id: `custom-logs-${uuidv4()}`,
+        type: 'logfile',
+        data_stream: {
+          namespace: 'default',
+        },
+        streams: [
+          {
+            id: `logs-onboarding-${pkgName}`,
+            data_stream: {
+              dataset: pkgName,
+            },
+            paths: integration.logFilePaths,
+          },
+        ],
+      };
+      try {
+        await packageClient.installCustomIntegration({
+          pkgName,
+          datasets: [{ name: pkgName, type: 'logs' }],
+        });
+        agentInputs.push(input);
+      } catch (error) {
+        // If the error is a naming collision, we can assume the integration is already installed and treat this step as successful
+        if (error instanceof NamingCollisionError) {
+          agentInputs.push(input);
+        } else {
+          throw error;
+        }
+      }
+    }
+  }
+  return agentInputs;
+}
+
+/**
+ * Parses and validates a TSV (tab-separated values) string of integrations with params.
+ *
+ * Returns an object of integrations to install.
+ *
+ * Example input:
+ *
+ * ```text
+ * system registry
+ * product_service custom /path/to/access.log
+ * product_service custom /path/to/error.log
+ * checkout_service custom /path/to/access.log
+ * checkout_service custom /path/to/error.log
+ * ```
+ */
+function parseIntegrationsTSV(tsv: string) {
```

Comment:
Seems a good function to write a unit test for, a few error scenarios to test

Reply:
Definitely agree, we will add an end to end test suite as a follow up.